### PR TITLE
Mapping the dynamic SensorUI to the sensors and its subscribers

### PR DESCRIPTION
**Pull request by jasmine-bae**
Working on creating a sensor UI that works in this way: When a drone is selected, the Sensor UI will showcase the name of a sensor on that drone, various toggles that switch on/off its subscribers, as well as "left" and "right" arrow buttons that cycle through each of the sensors on that specifically selected drone. 

_What has been done:_

- Moving the general architecture from ROSManager.cs to Drone.cs (under Drone.Select()). 

- Added new SensorManager.cs script that handles everything to do with the SensorUI. (Attached SensorUI gameobject in hierarchy)

- Created a Drone.selectedSensor public variable that will keep track of what sensor is currently being displayed (for that drone selected) on the sensorUI

- Finish up functions for the "left" and "right" arrow buttons. 

- Create toggles in the Unity Hierarchy

- Match toggles to its subscribers

- Cycle through sensors

- Added new tag "TOGGLESENSORS" which allows for deletion of previously created toggles when switching selected drones.

- Test if the sensors actually subscribe/unsubscribe from the toggles.

### CODE REVIEW

File: Assets/Scripts/Drone Scripts/Drone.cs [hunk 1]
```diff
@@ -207,6 +210,19 @@ public void DeleteWaypoint(Waypoint deletedWaypoint)
...
+            //Instantiating for sensor UI
+            //GameObject activeSensorName;
+            //Text senseTextComp;
+
+            //activeSensorName = GameObject.FindGameObjectWithTag("SENSORUI");
+            //senseTextComp = activeSensorName.GetComponent<Text>();
+
+            // We call a function on sensorUIManager -> Update UI (List<ROSSensorInterface>droneSensors)
+
+            SensorManager senseManage = new SensorManager();
```
Comment (PeruDayani) [inline Assets/Scripts/Drone Scripts/Drone.cs]: nitpicking and not necessary, but we can call the variable sensorManager for consistency?

File: Assets/Scripts/Drone Scripts/Drone.cs [hunk 2]
```diff
@@ -207,6 +210,19 @@ public void DeleteWaypoint(Waypoint deletedWaypoint)
...
+            senseManage.initializeSensorUI(attachedSensors);
```
Comment (PeruDayani) [inline Assets/Scripts/Drone Scripts/Drone.cs]: same

File: Assets/Scripts/Menu Scripts/SensorManager.cs [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+    // function 2: update the ui with the current selected sensor
+    // we get the list of subscribers: sensor.GetSensorSubscribers()
+    // this list of subscribers: eg: mesh, rad_level_1, rad_level3
+    // generate len(list) number of toggles  with the text of each being the subcsriber string
+    // upper bound of 6 subscibers per sensor
+
+    public void updateSensorUI(ROSSensorConnectionInterface inputSensor)
+    {
+
+        thisSensorText.text = inputSensor.ToString(); //where do we get the name of the sensor?
```
Comment (PeruDayani) [inline Assets/Scripts/Menu Scripts/SensorManager.cs]: Good point, I'll add a function in the interface

File: Assets/Scripts/Menu Scripts/SensorManager.cs [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+
+    // function 1: new drone selected: update all the sensors
+    // UpdateUI function: paramaters: List of Sensors
+    // store the list of sensors as current sensors
+    // selected sensor is the first sensor
+    // update the ui with the first sensor's subscribers.
+
+    public void initializeSensorUI(List<ROSSensorConnectionInterface> allSensors)
+    {
+        sensorList.AddRange(allSensors);
```
Comment (PeruDayani) [inline Assets/Scripts/Menu Scripts/SensorManager.cs]: I might be missing something obvious. 
But my understanding is that AddRange() appends onto the existing list. So if we switch drones the new list will contain the senors of the old drone + new drone?

File: Assets/Scripts/Drone Scripts/Drone.cs (hunk 1 above)
Comment (PeruDayani) [inline Assets/Scripts/Drone Scripts/Drone.cs]: why did we choose to create a new instance of sensorManger for every drone? 

we can have the sensorManager on the sensorUIGameobject and can reference it from WorldProperties maybe?

File: Assets/Scripts/Menu Scripts/SensorManager.cs [hunk 5]
```diff
@@ -36,6 +31,7 @@ void Awake()
     public void initializeSensorUI(List<ROSSensorConnectionInterface> allSensors)
     {
         sensorList.Clear();
+        deleteOldSensors();
```
Comment (PeruDayani) [inline Assets/Scripts/Menu Scripts/SensorManager.cs:37]: Do we need to delete the buttons and re create them?
We can just change the name of the sensors and remove/add as needed too?
But we don't have to do this.

File: Assets/Scripts/Drone Scripts/Drone.cs [hunk 6]
```diff
@@ -3,34 +3,42 @@
     using System.Collections;
     using System.Collections.Generic;
     using UnityEngine;
-    using ROSBridgeLib.interface_msgs;
+    using UnityEngine.UI;
+    using ROSBridgeLib.interface_msgs; //what is this for?
```
Comment (erwang01) [inline Assets/Scripts/Drone Scripts/Drone.cs]: Hm agreed. Lets open up an issue for this if no one knows.

File: Assets/Scripts/Drone Scripts/Drone.cs [hunk 7]
```diff
@@ -3,34 +3,42 @@
...
 
         public int nextWaypointId; // Incrementing counter to give all waypoints a unique ID when combined with the Drone ID
         public Dictionary<string, Waypoint> waypointsDict; // Collection of the waypoints in this drone's path
+
         public List<ROSSensorConnectionInterface> attachedSensors; // List of attached sensor gameobjects
 
         /// <summary>
         /// Constructor method for Drone class objects
         /// </summary>
         /// <param name="drone_obj"> We pass in a Gameobject for the drone -- this will be phased out and the new drone_obj gameObject will be instantiated in this method </param>
```
Comment (erwang01) [inline Assets/Scripts/Drone Scripts/Drone.cs:27]: Parameters no longer match documentation.

File: Assets/Scripts/Drone Scripts/Drone.cs [hunk 8]
```diff
@@ -175,38 +188,5 @@ public void DeleteWaypoint(Waypoint deletedWaypoint)
             Object.Destroy(deletedWaypoint.gameObjectPointer);
         }
 
-        /// <summary>
-        /// Use this to change which drone is selected in the world.
-        /// This also changes all drone aura materials so this drone is the only yellow one.
-        /// </summary>
-        public void Select() {
```
Comment (erwang01) [inline Assets/Scripts/Drone Scripts/Drone.cs:182]: So what is the replacement?

Comment (erwang01) [reply Assets/Scripts/Drone Scripts/Drone.cs:182]: jk found it in DroneProperties.

File: Assets/Scripts/Drone Scripts/DroneProperties.cs [hunk 9]
```diff
@@ -7,20 +7,54 @@
     public class DroneProperties : MonoBehaviour {
```
Comment (erwang01) [inline Assets/Scripts/Drone Scripts/DroneProperties.cs:7]: Can we have documentation for the class?

File: Assets/Scripts/Menu Scripts/SensorManager.cs [hunk 10]
```diff
@@ -0,0 +1,167 @@
+﻿using System.Collections;
+using System.Collections.Generic;
+using UnityEngine;
+using UnityEngine.UI;
+using ISAACS;
+
+public class SensorManager : MonoBehaviour {
```
Comment (erwang01) [inline Assets/Scripts/Menu Scripts/SensorManager.cs:10]: documentation for the class.

File: Assets/Scripts/ROS Scripts/ROS Connections v2/Sensors/LampSensor_ROSSensorConnection.cs [hunk 11]
```diff
@@ -77,6 +79,11 @@ void Update()
         }
     }
 
+    public string GetSensorName()
+    {
+        return sensorName;
```
Comment (erwang01) [inline Assets/Scripts/ROS Scripts/ROS Connections v2/Sensors/LampSensor_ROSSensorConnection.cs]: why not just return this.gameObject.name? That way you dont run the risk of someone modifying sensorName or gameObject.name without modifying the other?

File: Assets/Scripts/ROS Scripts/ROS Connections v2/Sensors/MeshSensor_ROSSensorConnection.cs [hunk 12]
```diff
@@ -148,6 +150,11 @@ void Update()
         }
     }
 
+    public string GetSensorName()
+    {
+        return sensorName;
```
Comment (erwang01) [inline Assets/Scripts/ROS Scripts/ROS Connections v2/Sensors/MeshSensor_ROSSensorConnection.cs]: ditto with the return this.gameObject.name;

you can add a setter if you want as well.

Also worth noting that idk if gameObjects can change their names...

File: Assets/Scripts/ROS Scripts/ROS Connections v2/Sensors/PCFaceSensor_ROSSensorConnection.cs [hunk 13]
```diff
@@ -68,6 +70,10 @@ void Update()
         }
     }
 
+    public string GetSensorName()
+    {
+        return sensorName;
```
Comment (erwang01) [inline Assets/Scripts/ROS Scripts/ROS Connections v2/Sensors/PCFaceSensor_ROSSensorConnection.cs]: Do you want to add this directly into the sensor interface? seems like you adding it for everyone.

File: Assets/Scripts/ROS Scripts/ROS Connections v2/Sensors/PointCloudSensor_ROSSensorConnection.cs [hunk 14]
```diff
@@ -50,6 +52,11 @@ void Update()
     }
 
 
+    public string GetSensorName()
+    {
+        return sensorName;
```
Comment (erwang01) [inline Assets/Scripts/ROS Scripts/ROS Connections v2/Sensors/PointCloudSensor_ROSSensorConnection.cs]: ditto to adding it to sensor interface.

File: Assets/Scripts/WorldProperties.cs [hunk 15]
```diff
@@ -92,22 +90,100 @@ void Start()
...
+
+            if (selectedDrone != null)
+            {
+                dronesQueue.Enqueue(selectedDrone);
+            }
+
+            if (dronesQueue.Count > 0)
             {
-                parent.GetComponent<Renderer>().material.shader = clipShader;
+                Drone nextDrone = dronesQueue.Dequeue();
```
Comment (erwang01) [inline Assets/Scripts/WorldProperties.cs:106]: do you not enque the drone back onto the end?

File: mono_crash.1979963d15.0.json [hunk 16]
```diff
@@ -0,0 +1,823 @@
+{
```
Comment (erwang01) [inline mono_crash.1979963d15.0.json]: Lets delete this file.

File: ProjectSettings/TagManager.asset [hunk 17]
```diff
@@ -36,6 +36,9 @@ TagManager:
   - LAMP
```
Comment (erwang01) [inline ProjectSettings/TagManager.asset:36]: What are tags?

**Review by erwang01 — DISMISSED** [review]
One file to delete a few comments but I think mostly non critical.

File: Assets/Scripts/Drone Scripts/DroneProperties.cs (hunk 9 above)
Comment (PeruDayani) [reply Assets/Scripts/Drone Scripts/DroneProperties.cs:7]: Done on the WaypointUpgrade branch.

File: Assets/Scripts/ROS Scripts/ROS Connections v2/Sensors/MeshSensor_ROSSensorConnection.cs (hunk 12 above)
Comment (jasmine-bae) [reply Assets/Scripts/ROS Scripts/ROS Connections v2/Sensors/MeshSensor_ROSSensorConnection.cs]: gameObjects can change their names btw

File: Assets/Scripts/WorldProperties.cs (hunk 15 above)
Comment (PeruDayani) [reply Assets/Scripts/WorldProperties.cs:106]: it's what line 99 does

File: ProjectSettings/TagManager.asset (hunk 17 above)
Comment (jasmine-bae) [reply ProjectSettings/TagManager.asset:36]: Tags can be attached to gameObjects to differentiate them from others - we use them in sensorManager to distinguish the subscriber toggles from any other gameobject in the scene.

**Review by exofunctor — APPROVED** [review]
looks good

**Pull request merged**